File: src/main/java/xyz/phanta/cbtweaker/multiblock/MultiBlockManager.java
package xyz.phanta.cbtweaker.multiblock;

import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.util.math.Vec2i;
import net.minecraft.util.ResourceLocation;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.ItemBlockRegistrar;
import xyz.phanta.cbtweaker.common.BlockMaterial;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;
import xyz.phanta.cbtweaker.gui.ScreenRegion;
import xyz.phanta.cbtweaker.gui.inventory.UiLayout;
import xyz.phanta.cbtweaker.recipe.RecipeLogic;
import xyz.phanta.cbtweaker.structure.StructureMatcher;
import xyz.phanta.cbtweaker.structure.StructureMatcherType;
import xyz.phanta.cbtweaker.util.ConfigException;
import xyz.phanta.cbtweaker.util.helper.DataLoadUtils;

import javax.annotation.Nullable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class MultiBlockManager {

    private static final UiLayout DEFAULT_UI_LAYOUT = new UiLayout(
            CbtTextureResources.GUI_GENERIC_SMALL, new ScreenRegion(7, 16, 162, 55), true, new Vec2i(7, 83), true);

    private final Path specDir;
    private final Map<String, JsonObject> mbSpecDtoCache = new HashMap<>();
    private final Map<String, MultiBlockType<?, ?, ?>> mbTypeTable = new HashMap<>();
    private final ItemBlockRegistrar registrar;

    public MultiBlockManager(Path specDir, ItemBlockRegistrar registrar) {
        this.specDir = specDir;
        this.registrar = registrar;
    }

    public void preloadAll() {
        CbtMod.LOGGER.info("Loading multi-block types...");
        try {
            Files.list(specDir).forEach(mbDir -> {
                try {
                    if (!Files.isDirectory(mbDir)) {
                        return;
                    }
                    Path specFile = mbDir.resolve("multiblock.json");
                    if (!Files.isRegularFile(specFile)) {
                        CbtMod.LOGGER.warn("Ignoring multi-block directory without a multiblock.json: {}", mbDir);
                        return;
                    }

                    JsonObject specDto = DataLoadUtils.readJsonFile(specFile).getAsJsonObject();
                    BlockMaterial blockMat = specDto.has("material")
                            ? BlockMaterial.fromString(specDto.get("material").getAsString()) : BlockMaterial.METAL;
                    RecipeLogic<?, ?, ?, ?, ?> recipeLogic = DataLoadUtils.loadRecipeLogic(
                            specDto.getAsJsonObject("recipe_logic"));
                    UiLayout uiLayout = specDto.has("ui")
                            ? CbtMod.PROXY.getTemplates().getUiTemplates().resolve(specDto.get("ui"))
                            : DEFAULT_UI_LAYOUT;

                    MultiBlockType<?, ?, ?> mbType = new MultiBlockType<>(
                            mbDir.getFileName().toString(), blockMat, mbDir, recipeLogic, uiLayout);
                    mbSpecDtoCache.put(mbType.getId(), specDto);
                    mbTypeTable.put(mbType.getId(), mbType);
                    registrar.addBlockRegistration("mb_" + mbType.getId(), new MultiBlockControllerBlock(mbType));
                    CbtMod.LOGGER.debug("Loaded multi-block: {}", mbType.getId());
                } catch (Exception e) {
                    throw new IllegalStateException("Failed to load multi-block specification: " + mbDir, e);
                }
            });
        } catch (Exception e) {
            throw new IllegalStateException("Failed to load multi-block types!", e);
        }
        CbtMod.LOGGER.info("Finished loading multi-block specifications.");
    }

    public void loadAll() {
        CbtMod.LOGGER.info("Loading multi-block data...");
        for (MultiBlockType<?, ?, ?> mbType : mbTypeTable.values()) {
            try {
                JsonObject specDto = mbSpecDtoCache.get(mbType.getId());

                JsonObject structDto = specDto.getAsJsonObject("structure");
                ResourceLocation structTypeId = new ResourceLocation(structDto.get("type").getAsString());
                StructureMatcherType structType = CbtMod.PROXY.getStructureMatchers()
                        .lookUpStructureMatcher(structTypeId);
                if (structType == null) {
                    throw new ConfigException("Unknown structure matcher type: " + structTypeId);
                }
                StructureMatcher struct = structType.loadMatcher(mbType, structDto);

                lateInitMultiBlockType(mbType, struct, specDto.getAsJsonObject("recipe_logic"));
                CbtMod.LOGGER.debug("Initialized multi-block: {}", mbType.getId());
            } catch (Exception e) {
                throw new IllegalStateException("Failed to load data for multi-block: " + mbType.getId(), e);
            }
        }
        mbSpecDtoCache.clear();

        CbtMod.LOGGER.info("Loading multi-block recipes...");
        for (MultiBlockType<?, ?, ?> mbType : mbTypeTable.values()) {
            loadRecipesForMultiBlock(mbType);
        }
        CbtMod.LOGGER.info("Finished loading multi-block data.");
    }

    private static <R, C, D> void lateInitMultiBlockType(MultiBlockType<R, C, D> mbType,
                                                         StructureMatcher struct, JsonObject recipeLogicDto) {
        RecipeLogic<R, C, D, ?, ?> recipeLogic = mbType.getRecipeLogic();
        C recipeConfig = recipeLogic.loadConfig(recipeLogicDto);
        mbType.init(struct, recipeConfig, recipeLogic.createEmptyDatabase(recipeConfig));
    }

    private <R, C, D> void loadRecipesForMultiBlock(MultiBlockType<R, C, D> mbType) {
        try {
            DataLoadUtils.loadRecipes(mbType.getId(), mbType.getRecipeDirectory(), "multiblock.json",
                    mbType.getRecipeLogic(), mbType.getRecipeConfig(), mbType.getRecipeDatabase());
        } catch (Exception e) {
            throw new IllegalStateException("Failed to load recipes for multi-block: " + mbType.getId());
        }
    }

    public Collection<MultiBlockType<?, ?, ?>> getAll() {
        return mbTypeTable.values();
    }

    @Nullable
    public MultiBlockType<?, ?, ?> lookUp(String id) {
        return mbTypeTable.get(id);
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/multiblock/MultiBlockManager.kt
package st.evening.mc.cbtweaker.multiblock

import net.minecraft.util.ResourceLocation
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.common.BlockConfig
import st.evening.mc.cbtweaker.gui.inventory.WindowConfig
import st.evening.mc.cbtweaker.util.DataGenHelper
import st.evening.mc.cbtweaker.util.forEachFile
import st.evening.mc.prelude.api.data.ser.SerializationException
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.data.tjson.TypedJsonParser
import st.evening.mc.prelude.api.data.tjson.useAny
import st.evening.mc.prelude.api.data.tjson.useObject
import st.evening.mc.prelude.api.data.tjson.useObjectValue
import st.evening.mc.prelude.api.data.tjson.useStringValue
import st.evening.mc.prelude.api.registration.ModRegistrar
import java.nio.file.Files
import java.nio.file.Path
import kotlin.io.path.readText

class MultiBlockManager(private val specDir: Path, private val reg: ModRegistrar) : Iterable<MultiBlockType<*>> {
    private val mbSpecDtoCache: MutableMap<String, TJson.Object> = mutableMapOf()
    private val mbTypeTable: MutableMap<String, MultiBlockType<*>> = mutableMapOf()

    operator fun get(id: String): MultiBlockType<*>? = mbTypeTable[id]

    override fun iterator(): Iterator<MultiBlockType<*>> = mbTypeTable.values.iterator()

    fun preloadAll() {
        CbTweaker.logger.info("Loading multi-block types...")
        try {
            specDir.forEachFile { specFile ->
                try {
                    if (!Files.isRegularFile(specFile)) return@forEachFile
                    val specFileName = specFile.fileName.toString()
                    if (!specFileName.endsWith(".tjson")) {
                        CbTweaker.logger.warn(
                            "Ignoring non-TJSON file in multi-block specification directory: {}",
                            specFileName
                        )
                        return@forEachFile
                    }
                    val mbId = specFileName.dropLast(6)
                    if (mbId.isBlank()) {
                        throw SerializationException("Empty multi-block ID!")
                    }

                    val specDto = TypedJsonParser.parseObject(specFile.readText())
                    val mbType = JsonPath.atRoot {
                        MultiBlockType(
                            reg,
                            mbId,
                            specDto.useObject("block") { BlockConfig.load(it) } ?: BlockConfig.DEFAULT,
                            specDto.useObjectValue("behaviour") { behaviourDto ->
                                behaviourDto.useStringValue("type") {
                                    CbTweaker.defns.machineBehaviours[ResourceLocation(it)]
                                        ?: throw SerializationException.withPath("Unknown machine behaviour: $it")
                                }
                            },
                            specDto.useAny("ui") {
                                CbTweaker.defns.templates.windowTemplates.resolve(it)
                            } ?: WindowConfig.GENERIC_SMALL
                        )
                    }
                    mbSpecDtoCache[mbId] = specDto
                    mbTypeTable[mbId] = mbType
                    CbTweaker.logger.debug("Loaded multi-block: {}", mbId)
                } catch (e: Exception) {
                    throw IllegalStateException("Failed to load multi-block specification: $specFile", e)
                }
            }
        } catch (e: Exception) {
            throw IllegalStateException("Failed to load multi-block types!", e)
        }
        CbTweaker.logger.info("Finished loading multi-block specifications.")
    }

    fun dataGenBlockModels(resourceDir: Path) {
        if (mbTypeTable.isEmpty()) return
        val cbtDir = resourceDir.resolve("cbtweaker")
        val blockStateDir = Files.createDirectories(cbtDir.resolve("blockstates"))
        val itemModelDir = Files.createDirectories(cbtDir.resolve("models/item"))
        mbTypeTable.keys.forEach { id ->
            val bsFile = blockStateDir.resolve("mb_$id.json")
            if (!Files.exists(bsFile)) {
                CbTweaker.logger.info("Generating multi-block controller block state mapping: ${bsFile.fileName}")
                DataGenHelper.writeToFile(bsFile, DataGenHelper.machineBlockState)
            }
            val imFile = itemModelDir.resolve("mb_$id.json")
            if (!Files.exists(imFile)) {
                CbTweaker.logger.info("Generating multi-block controller item model: ${imFile.fileName}")
                DataGenHelper.writeToFile(imFile, DataGenHelper.machineItemModel)
            }
        }
    }

    fun loadAll() {
        CbTweaker.logger.info("Loading multi-block data...")
        mbTypeTable.forEach { (mbId, mbType) ->
            try {
                lateInitMultiBlockType(mbType, mbSpecDtoCache.getValue(mbId))
                CbTweaker.logger.debug("Initialized multi-block: {}", mbId)
            } catch (e: Exception) {
                throw IllegalStateException("Failed to load data for multi-block: $mbId", e)
            }
        }
        mbSpecDtoCache.clear()
    }

    private fun <S> lateInitMultiBlockType(mbType: MultiBlockType<S>, specDto: TJson.Object) {
        JsonPath.atRoot {
            mbType.init(
                specDto.useObjectValue("structure") { structDto ->
                    structDto.useStringValue("type") {
                        CbTweaker.defns.structureMatchers[ResourceLocation(it)]
                            ?: throw SerializationException.withPath("Unknown structure matcher type: $it")
                    }.loadMatcher(mbType, structDto)
                },
                specDto.useObjectValue("behaviour") {
                    mbType.behaviour.loadStateFactory(mbType, it)
                }
            )
        }
    }
}
